service/manager.go

```diff
@@ -330,29 +330,28 @@ func (m *manager) start() Control {
 		errs := m.startModules()
 		m.registerSignalHandlers()
 		if len(errs) > 0 {
-			var serviceErr error
-			errChan := make(chan Exit, 1)
-			// grab the first error, shut down the service and return the error
+			// shut down the service using the first error
+			serviceErr := errs[0]
+			m.shutdownMu.Unlock()
+
+			if _, err := m.shutdown(serviceErr, "", nil); err != nil {
+				zap.L().Error("Unable to shut down modules",
+					zap.NamedError("initialError", serviceErr),
+					zap.NamedError("shutdownError", err),
+				)
+			}
+
+			// emit all errors over the errChan channel for logging purposes
+			errChan := make(chan Exit, len(errs))
```

Comment:
I think we can actually make this quite a bit simpler: rather than attempting to send separate errors for each failing module, we can compose all failures into a single error. That should eliminate most of the complexity in handling channels and mutexes.

Use `go.uber.org/multierr`, and try something like this:

```
if len(errs) > 0 {
  combined := multierr.Combine(errs...)
  errChan := make(chan Exit, 1)
  errChan <- Exit{ ...stuff... }
  zap.L().Error("Error starting modules", zap.Error(combined))
  m.shutdownMu.Unlock()
  return Control{...stuff...}
}
```

Since that brings some sanity to the mutex handling (lock once at the beginning of the method, unlock before every return), we can remove the duplicated unlocks and change the beginning of the method to

```
m.shutdownMu.Lock()
defer m.shutdownMu.Unlock()
```

Reply:
Thanks for introducing me to go.uber.org/multierr, its great! Cleaned up the code as you suggested, so much easier to read now :)